Add Configuration Builder and Better Download page

joshuaogle opened the pull request:
- Make Download Link more prominent on home page (like new Consul site)
- Add UI Demo link to home page
- Download page now suggests download based on your current system
- Added links for next steps

This came out of the research for UI Setup Wizards, and realizing that our site could do a few things to make setup easier. I included these website updates in my prototype for Setup Wizards and was asked to do the development too since it wasn't too complicated and the Vault front-end team are busy on 0.11 features. This isn't closely related to the Setup Wizards (just in the same spirit) so it can go live separately.

I've tested it well but it could definitely use some additional eyes on it to make sure it's a great experience. Thanks!

@jescalan I'm not sure you've seen any of this yet, wanted to make sure you were aware. If you have any Marketing related feedback (or any other kind) I'd love to hear it.

@mitchellh I think we may have skipped this bit when reviewing the Setup Wizard prototype, so just a heads up for you too since you've put a lot of thought into how to make these parts better.

# Home page
### Before:
![image](https://user-images.githubusercontent.com/111194/43668587-fa49541a-973a-11e8-9faf-af9bbd41ceca.png)

### After:
![image](https://user-images.githubusercontent.com/111194/43668580-f34fed54-973a-11e8-96f9-ee58d6b2a38b.png)
---

# Download suggestion
### Before:
![image](https://user-images.githubusercontent.com/111194/43668673-5aef2254-973b-11e8-9ff1-51df2d8a7b1f.png)

### After:
![image](https://user-images.githubusercontent.com/111194/43668697-7b8b4704-973b-11e8-984b-728da0a4b85e.png)

---

# Download next steps
![image](https://user-images.githubusercontent.com/111194/43668717-8ff31dde-973b-11e8-886b-4c4fb860d9ce.png)

---

# Configuration Builder
![image](https://user-images.githubusercontent.com/111194/43668761-cc0d5d2a-973b-11e8-9d8a-2e6ba82bc8d6.png)

Some of these fields aren't things we want the user putting into a web form, so we disable the field and put a placeholder in the config file for them to replace later.
![image](https://user-images.githubusercontent.com/111194/43668881-9f8b51a2-973c-11e8-9732-c6efbde05123.png)

---

# Starting the Server
If they choose to add the UI flag in the builder, they'll go to this tab on the next step, which gets them into the UI sooner and will lead to the Setup Wizard in the UI. If they don't, they'll get the normal CLI instructions instead.
![image](https://user-images.githubusercontent.com/111194/43668775-e01460d4-973b-11e8-81df-797739c1b961.png)

meirish commented on website/source/assets/javascripts/configuration-builder.js:
```diff
@@ -0,0 +1,133 @@
+
+  // Add Listener stanza
+  if (document.getElementById("include_tcp_listener").checked) {
+    config += `listener "tcp" {
```
I don't know enough about the compilation here, but does middleman transpile this to ES5? Or will this only be supported in browsers that support ES6 template strings?

jescalan replied:
This middleman setup does not transpile es6, so we need to have all es5 javascript, unfortunately

joshuaogle replied:
It does, yeah. I thought I remembered IE11 support at least but looks like it might be a good idea to do it the old way
https://caniuse.com/#feat=template-literals

meirish replied:
(looks like it's just IE that doesn't have native support for this fwiw)

meirish replied:
Lol I didn't refresh and both of you had already replied 😂 sorry bout that.

meirish commented on website/source/assets/javascripts/configuration-builder.js:
```diff
@@ -0,0 +1,133 @@
+  saveAs(blob, "vault-config.hcl");
+  document.querySelector(".form-actions").style.display = "none";
+  document.querySelector("#download-confirm").style.display = "block";
+  return false;
```
Just curious, is the `return false` important?

joshuaogle replied:
It was at one point because clicking the button was refreshing the page and keeping you from seeing the confirmation message. I ended up having to add `type="button"` for another bug and looks like it doesn't refresh now (should have realized why that was happening earlier). Thanks!

meirish commented on website/source/assets/stylesheets/_config.scss:
```diff
@@ -0,0 +1,64 @@
+  align-items: center;
+  color: $sidebar-link-color-active;
+  cursor: pointer;
+  display: flex;
```
Any IE boogeymen to look out for here? (I think mostly no since there's not `flex:` declarations).

joshuaogle replied:
That's okay. If it can't flex then it will be full width like it is currently

jescalan replied:
At least from the digital team perspective, we do not support browsers that don't support flexbox, so this is 👍 for me haha

meirish commented on website/source/docs/configuration/builder.html.erb:
```diff
@@ -0,0 +1,70 @@
+    name: "ui",
+    options: {
+      false: "Do not activate UI",
+      true: "Activate UI"
```
Do we want Activate to be the default here since the option would be left out if they don't check, but also if they do check and then don't switch the <select>?

jescalan commented on website/source/assets/javascripts/configuration-builder.js:
```diff
@@ -0,0 +1,119 @@
+    }
+  }
+  return;
+}
```
This is very much an optional, opinionated thing, but I typically favor placing utility functions at the bottom of the file, rather than the top. Since javascript will hoist them anyway, this won't cause any sort of ordering error, and it makes it more clear for others to read the implementation portion of the code, scrolling down to reference the utility functions as needed, then returning to the top of the document to resume the code reading, than scrolling through out-of-context utility functions, then going back up and down to an arbitrary spot in the middle to get back to the reading.

jescalan commented on website/source/assets/javascripts/configuration-builder.js:
```diff
@@ -0,0 +1,119 @@
+        document.querySelector('[data-if-option="' + selection + '"]').classList.toggle("active");
+      }
+    });
+  }
```
I might have done with `revealTriggers.map` and `configTriggers.map` here to make the code a little easier to read, but of course this is just a stylistic thing.

jescalan commented on website/source/assets/javascripts/downloads.js:
```diff
@@ -0,0 +1,22 @@
+
+  for (i = 0; i < downloadLinks.length; i++) {
+    downloadLinks[i].addEventListener("click", handleDownloadLinkClick);
+  }
```
Same here with the `map`

jescalan commented on website/source/assets/javascripts/downloads.js:
```diff
@@ -0,0 +1,22 @@
+    confirm.querySelector(".chosen-os-name").innerHTML = name;
+    confirm.querySelector(".chosen-os-bit").innerHTML = bit;
+    confirm.querySelector(".icon").innerHTML = icon;
+  }
```
In the other files, the function has sat up top in the global scope, but here it's nested within the turbolinks load block. Any reason for this?

jescalan commented:
### Some UI Feedback

![](https://cl.ly/2l1I403Y1R3f/Image%202018-08-06%20at%2012.40.39%20PM.png)

![](https://cl.ly/2y3J0z2r3d2q/Image%202018-08-06%20at%2012.42.39%20PM.png)

![](https://cl.ly/2W2A0s0Q3A1q/Image%202018-08-06%20at%2012.46.03%20PM.png)

👏 overall this is very cool!

joshuaogle commented:
@jescalan Thanks, that's all great feedback!

I refactored the JS and moved those functions. Good call about `map` (or in this case `forEach`) being easier to read.

The demo links were aligned right on their line but you're right that it looked odd not lining up with the buttons above them. I added "Vault Web UI" which made them line up better.

![image](https://user-images.githubusercontent.com/111194/43734522-3d2a8c9a-9975-11e8-93f2-a1adeaf89074.png)

The download links were clearer in my original design but I was hoping it would be good enough just making that link a button and not moving it to the front of the list. I agree it wasn't clear though. Updated to the original design like so:

![image](https://user-images.githubusercontent.com/111194/43734440-0105f3d0-9975-11e8-8005-a563f8c4e7bb.png)

jescalan commented:
Those two design solutions look tons better - thanks for taking the time to go through this! Really a fan of the download button change there, it's way more clear 😁
